datafusion/core/src/dataframe/parquet.rs

```diff
@@ -246,4 +249,136 @@ mod tests {
+        let mut writer = ArrowWriter::try_new(file, Arc::clone(&schema), Some(props))?;
+
+        writer.write(&first_batch)?;
+        let mut num_rows = first_batch.num_rows();
+
+        for batch in batches {
+            writer.write(&batch)?;
+            num_rows += batch.num_rows();
+        }
+        writer.close()?;
+        Ok(num_rows)
+    }
+
+    #[tokio::test]
+    async fn roundtrip_parquet_with_encryption() -> Result<()> {
+        use parquet::encryption::decrypt::FileDecryptionProperties;
+        use parquet::encryption::encrypt::FileEncryptionProperties;
+        use datafusion_common::config::ConfigFileDecryptionProperties;
+        use crate::execution::SessionStateBuilder;
+        
+        let test_df = test_util::test_table().await?;
+
+        let schema = test_df.schema();
+        let footer_key = b"0123456789012345".to_vec(); // 128bit/16
+        let column_key = b"1234567890123450".to_vec(); // 128bit/16
+
+        let mut encrypt = FileEncryptionProperties::builder(footer_key.clone());
+        let mut decrypt = FileDecryptionProperties::builder(footer_key.clone());
+       
+        for field in schema.fields().iter() {
+            encrypt = encrypt.with_column_key(field.name().as_str(), column_key.clone());
+            decrypt = decrypt.with_column_key(field.name().as_str(), column_key.clone());
+        }
+        
+        let encrypt = encrypt.build().unwrap();
+        let decrypt = decrypt.build().unwrap();
+        
+        let df = test_df.clone();
+        let tmp_dir = TempDir::new()?;
+        let tempfile = tmp_dir.path().join("roundtrip.parquet");
+        let output_path = "file://local/roundtrip.parquet";
+        
+        let num_rows_written: usize = if true {
+            // Write encrypted parquet using ArrowWriter
+            let props = WriterProperties::builder()
+                .with_file_encryption_properties(encrypt)
+                .build();
+
+            let batches = df.collect().await?;
+            write_batches(tempfile.clone(), props, batches).unwrap()
+        } else {
+            // Write encrypted parquet using write_parquet
+            let local = Arc::new(LocalFileSystem::new_with_prefix(&tmp_dir)?);
+            let local_url = Url::parse("file://local").unwrap();
+            let ctx = &test_df.session_state;
+            ctx.runtime_env().register_object_store(&local_url, local);
+            let mut options = TableParquetOptions::default();
+            options.global.file_encryption_properties = Some(encrypt.clone().into());
+            df.write_parquet(
+                output_path,
+                DataFrameWriteOptions::new().with_single_file_output(true),
+                Some(options),
+            )
+                .await?;
+            test_df.count().await?
+        };
+        
+        
+        // Check that file actually used encryption
+        // TODO, I guess arrow-rs does not have a serialized file reader that supports encryption?
```

Comment:
Yeah I don't think this is currently implemented, you have to use the `ParquetRecordBatchReaderBuilder` type for reading. It looks like there are a lot of different APIs for reading Parquet in arrow-rs... Is this a blocker for the DataFusion implementation or it would just be useful for testing?

Reply:
So, my original goal here was to follow the conventions in the other tests. I think it is actually possible to create this since, e.g. the batch reader somehow uses a SerializedFileReader underneath so I wanted to hit you up to see if you had an example handy. Anyway, I will probably go in a different direction and actually use a DataFrame read here since what I really want to test is the higher level API. For now, I am still stuck on tracking a hideous bug as noted above.